zebra/zebra_rnh.c

```diff
@@ -623,10 +628,34 @@ zebra_rnh_resolve_nexthop_entry(struct zebra_vrf *zvrf, afi_t afi,
 				break;
 		}
 
-		/* Route entry found, we're done; else, walk up the tree. */
+		/* Route entry found, we're done; else, walk up the tree.
+		 *
+		 * BGP routes are usually recursive, though. Keep resolving
+		 * on the nexthop until hitting a non-BGP route.
+		 */
 		if (re) {
-			*prn = rn;
-			return re;
+			if (re->type != ZEBRA_ROUTE_BGP) {
+				*prn = rn;
+				return re;
+			}
+
+			/*
+			 * Note an aggregate has the NEXTHOP_TYPE_BLACKHOLE
+			 * as the nexthop type.
+			 */
+			nexthop = re->nhe->nhg.nexthop;
+			if (!nexthop || ((nexthop->type != NEXTHOP_TYPE_IPV4) &&
```

Comment:
it's possible that a NEXTHOP_TYPE_IPV4 will be fully resolved (I know).  Please just check the nexthop->resolved pointer for this test.

Reply:
Let me investigate further

Reply:
I missed this comment. Will do.

Reply:
Done in the next revision